File: sdk/py/hop_top_kit/telemetry/consent.py
"""Consent-file reader.

Reads the persisted consent decision from the kit AppConfig at
``<XDG_CONFIG_HOME>/kit/config.yaml`` under the ``kit.telemetry.consent``
partition. A pre-refactor layout at ``<XDG_CONFIG_HOME>/kit/telemetry.yaml``
(bare ``telemetry.consent``) is read as a fallback for backward
compatibility with installs that have not been migrated yet.

Default-denied: any error, missing file, malformed YAML, or unknown
state value yields ``Consent.denied()``.

The Go-side ``core/consent/file_store.go`` owns the persisted schema:

    kit:
      telemetry:
        consent:
          state: granted | denied
          decided_at: RFC3339
          prompt_version: int
          decision_source: prompt | flag | env | config

SDKs only read; they never write this file.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

import yaml


@dataclass(frozen=True)
class Consent:
    """In-memory representation of the persisted consent decision.

    The persisted file does NOT carry mode; that's env/context-only.
    """

    allowed: bool
    prompt_version: int = 0
    decision_source: str = "config"
    decided_at: Optional[str] = None

    @classmethod
    def denied(cls) -> Consent:
        """Default-denied sentinel for missing/malformed input."""
        return cls(allowed=False)


def _xdg_config_home() -> Path:
    return Path(os.environ.get("XDG_CONFIG_HOME") or (Path.home() / ".config"))


def consent_path() -> Path:
    """Canonical path for the persisted consent YAML (config.yaml)."""
    return _xdg_config_home() / "kit" / "config.yaml"


def legacy_consent_path() -> Path:
    """Pre-refactor consent file path (telemetry.yaml).

    Read-only fallback used by :func:`load_consent` when the canonical
    config.yaml is missing or lacks the ``kit.telemetry.consent`` block.
    """
    return _xdg_config_home() / "kit" / "telemetry.yaml"


def _consent_from_block(block: object) -> Optional[Consent]:
    """Decode a consent: mapping into a Consent or None if unusable."""
    if not isinstance(block, dict):
        return None

    state = block.get("state", "denied")
    if state not in ("granted", "denied"):
        return None

    try:
        prompt_version = int(block.get("prompt_version", 0))
    except (TypeError, ValueError):
        prompt_version = 0

    decision_source = str(block.get("decision_source", "config"))
    decided_at = block.get("decided_at")
    if decided_at is not None and not isinstance(decided_at, str):
        decided_at = str(decided_at)

    return Consent(
        allowed=(state == "granted"),
        prompt_version=prompt_version,
        decision_source=decision_source,
        decided_at=decided_at,
    )


def _load_yaml(path: Path) -> Optional[dict]:
    if not path.exists():
        return None
    try:
        data = yaml.safe_load(path.read_text())
    except Exception:
        return None
    return data if isinstance(data, dict) else None


def load_consent() -> Consent:
    """Read the consent decision. Never raises.

    Tries the canonical ``config.yaml`` (``kit.telemetry.consent``)
    first; falls back to the legacy ``telemetry.yaml``
    (``telemetry.consent``) when the canonical file is absent or lacks
    the consent block. Missing files, malformed YAML, unexpected types,
    or unknown ``state`` values all map to ``Consent.denied()``.
    """
    # Canonical path: <XDG_CONFIG_HOME>/kit/config.yaml under
    # kit.telemetry.consent.
    data = _load_yaml(consent_path())
    if data is not None:
        kit = data.get("kit")
        if isinstance(kit, dict):
            tel = kit.get("telemetry")
            if isinstance(tel, dict):
                got = _consent_from_block(tel.get("consent"))
                if got is not None:
                    return got

    # Legacy fallback: <XDG_CONFIG_HOME>/kit/telemetry.yaml under
    # bare telemetry.consent.
    legacy = _load_yaml(legacy_consent_path())
    if legacy is not None:
        tel = legacy.get("telemetry")
        if isinstance(tel, dict):
            got = _consent_from_block(tel.get("consent"))
            if got is not None:
                return got

    return Consent.denied()

File: sdk/ts/src/telemetry/consent.ts
/**
 * @module telemetry/consent
 * @package @hop-top/kit
 *
 * Read-only consent-file loader — the TS mirror of the read path in
 * `go/core/consent/file_store.go`.
 *
 * On-disk shape (YAML under `<XDG_CONFIG_HOME>/kit/config.yaml` at the
 * `kit.telemetry.consent` partition):
 *
 * ```yaml
 * kit:
 *   telemetry:
 *     consent:
 *       state: granted          # | denied
 *       prompt_version: 1
 *       decision_source: prompt # | flag | config | env
 *       decided_at: 2026-05-19T12:00:00Z
 * ```
 *
 * A pre-refactor layout at `<XDG_CONFIG_HOME>/kit/telemetry.yaml` (with
 * a bare `telemetry.consent` block at the top level) is read as a
 * fallback so installs that have not been migrated yet still work.
 *
 * This TS module is **read-only**: writing the consent file remains the
 * job of the Go `kit consent` CLI. SDK consumers just need to know
 * whether emission is allowed.
 *
 * Anything that isn't a clean, parseable, `state: granted` block
 * collapses to `deniedConsent` — a fail-closed default so a corrupt
 * config can never accidentally enable emission.
 */

import { promises as fs } from 'node:fs';
import { existsSync } from 'node:fs';
import { homedir } from 'node:os';
import * as path from 'node:path';
import * as yaml from 'js-yaml';

/**
 * Consent is the resolved telemetry decision for the current
 * installation. The shape mirrors the relevant subset of the Go
 * `consent.Decision` struct.
 */
export interface Consent {
  /** allowed is true only when state == "granted". */
  readonly allowed: boolean;
  /** promptVersion records which prompt revision the user accepted. */
  readonly promptVersion: number;
  /**
   * decisionSource records how the decision was captured —
   * "prompt" | "flag" | "config" | "env". Defaults to "config" when
   * the file is missing or unreadable.
   */
  readonly decisionSource: string;
  /** decidedAt is the RFC3339 timestamp of the decision, if present. */
  readonly decidedAt?: string;
}

/**
 * deniedConsent is the canonical fail-closed default. Returned whenever
 * the file is missing, unreadable, malformed, or carries an unknown
 * state value.
 */
export const deniedConsent: Consent = {
  allowed: false,
  promptVersion: 0,
  decisionSource: 'config',
};

/** xdgConfigHome returns `$XDG_CONFIG_HOME` or `~/.config`. */
function xdgConfigHome(): string {
  return process.env.XDG_CONFIG_HOME ?? path.join(homedir(), '.config');
}

/**
 * consentPath returns the canonical on-disk path used by this module
 * (`<XDG_CONFIG_HOME>/kit/config.yaml`). Useful for `kit telemetry
 * status` diagnostics and SDK consumers that want to surface the
 * location.
 */
export function consentPath(): string {
  return path.join(xdgConfigHome(), 'kit', 'config.yaml');
}

/**
 * legacyConsentPath returns the pre-refactor consent file location
 * (`<XDG_CONFIG_HOME>/kit/telemetry.yaml`). Read-only fallback consumed
 * by `loadConsent`; SDK callers should prefer `consentPath`.
 */
export function legacyConsentPath(): string {
  return path.join(xdgConfigHome(), 'kit', 'telemetry.yaml');
}

/**
 * loadConsent reads and parses the consent file, returning a fully
 * resolved Consent struct. Failure modes fold to `deniedConsent`:
 *
 *   - File missing → denied.
 *   - File unreadable → denied.
 *   - YAML parse error → denied.
 *   - Missing `kit.telemetry.consent` block (and no legacy) → denied.
 *   - `state` neither "granted" nor "denied" → denied.
 *
 * `state: "denied"` returns a Consent with `allowed=false` but with the
 * other fields populated, so callers can distinguish an explicit deny
 * from a fail-closed default.
 *
 * Read order: canonical `config.yaml` (`kit.telemetry.consent`) first,
 * then legacy `telemetry.yaml` (`telemetry.consent`) as a fallback.
 */
export async function loadConsent(): Promise<Consent> {
  const canonical = await readConsentBlock(consentPath(), 'canonical');
  if (canonical !== null) {
    return canonical;
  }
  const legacy = await readConsentBlock(legacyConsentPath(), 'legacy');
  if (legacy !== null) {
    return legacy;
  }
  return deniedConsent;
}

/**
 * readConsentBlock loads one YAML file and tries to decode a consent
 * block from it. Returns the resolved Consent, or null when the file
 * is missing / unreadable / malformed / lacks the expected block.
 * Distinguishing null from `deniedConsent` lets the caller fall
 * through to the next candidate path.
 */
async function readConsentBlock(
  p: string,
  variant: 'canonical' | 'legacy',
): Promise<Consent | null> {
  if (!existsSync(p)) {
    return null;
  }

  let data: unknown;
  try {
    const raw = await fs.readFile(p, 'utf8');
    data = yaml.load(raw);
  } catch {
    return null;
  }

  if (typeof data !== 'object' || data === null) {
    return null;
  }
  const root = data as Record<string, unknown>;

  // Canonical: kit.telemetry.consent. Legacy: telemetry.consent (no kit).
  let telemetry: unknown;
  if (variant === 'canonical') {
    const kit = root.kit;
    if (typeof kit !== 'object' || kit === null) {
      return null;
    }
    telemetry = (kit as Record<string, unknown>).telemetry;
  } else {
    telemetry = root.telemetry;
  }

  if (typeof telemetry !== 'object' || telemetry === null) {
    return null;
  }
  const block = (telemetry as Record<string, unknown>).consent;
  if (typeof block !== 'object' || block === null) {
    return null;
  }
  const c = block as Record<string, unknown>;

  const state = c.state;
  if (state !== 'granted' && state !== 'denied') {
    return null;
  }

  return {
    allowed: state === 'granted',
    promptVersion: typeof c.prompt_version === 'number' ? c.prompt_version : 0,
    decisionSource:
      typeof c.decision_source === 'string' ? c.decision_source : 'config',
    decidedAt: parseDecidedAt(c.decided_at),
  };
}

/**
 * parseDecidedAt coerces the raw YAML value into an RFC3339 string.
 * js-yaml's default schema promotes unquoted ISO timestamps to JS
 * `Date` objects, so we accept either form. Go writes `time.Time` as
 * an RFC3339 string with the YAML `!!timestamp` tag — both reader
 * paths converge here.
 */
function parseDecidedAt(v: unknown): string | undefined {
  if (typeof v === 'string') {
    return v;
  }
  if (v instanceof Date && !Number.isNaN(v.getTime())) {
    return v.toISOString().replace(/\.\d{3}Z$/, 'Z');
  }
  return undefined;
}
